cloudant-sync-datastore-core/src/main/java/com/cloudant/mazha/CouchClient.java

```diff
@@ -91,32 +165,35 @@ private InputStream executeToInputStream(HttpConnection connection) throws Couch
         // responses (eg 404 throws a FileNotFoundException) but we need to map to our own
         // specific exceptions
         try {
-            is = connection.execute().responseAsInputStream();
+            inputStream = connection.execute().responseAsInputStream();
         } catch (IOException ioe) {
+            System.out.println("*** cause "+ioe);
             cause = ioe;
         }
 
+        // response code and message will generally be present together or not all
         try {
-            code = connection.getConnection().getResponseCode();
-            response = connection.getConnection().getResponseMessage();
-            // everything ok? return the stream
-            if (code / 100 == 2) { // success [200,299]
-                return is;
-            } else if (code == 404) {
-                throw new NoResourceException(response, cause);
-            } else {
-                es = connection.getConnection().getErrorStream();
-                // TODO what if deserialisation fails?
-                CouchException ex = this.jsonHelper.fromJson(new InputStreamReader(es), CouchException.class);
-                ex.setStatusCode(code);
-                throw ex;
-            }
+            responseCode = connection.getConnection().getResponseCode();
+            responseMessage = connection.getConnection().getResponseMessage();
         } catch (IOException ioe) {
-            throw new CouchException("Error retrieving server response", ioe, code);
+            responseMessage = "Error retrieving server response message";
+        }
+
+        // error stream will be present or null if not applicable
+        errorStream = connection.getConnection().getErrorStream();
+
+        try {
+            ExecuteResult executeResult = new ExecuteResult(inputStream,
+                    errorStream,
+                    responseCode,
+                    responseMessage,
+                    cause,
+                    jsonHelper);
+            return executeResult;
         } finally {
-            if (es != null) {
+            if (errorStream != null) {
```

Comment:
Unfair to call this up when all you did was rename the field to something that made more sense, but wouldn't this be a good place for `IOUtils.close()`?